src/box/iproto.cc

```diff
@@ -2447,37 +2515,57 @@ iproto_send_stop_msg(void)
 }
 
 static inline int
-iproto_send_listen_msg(struct evio_service *binary)
+iproto_send_listen_msg(struct evio_service *binaries, uint32_t binaries_count)
 {
 	struct iproto_cfg_msg cfg_msg;
 	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_LISTEN);
-	cfg_msg.binary = binary;
+	assert(binaries_count <= IPROTO_LISTEN_SOCKET_MAX);
+	for (uint32_t i = 0; i < binaries_count; i++)
+		cfg_msg.binaries[i] = &binaries[i];
+	cfg_msg.binaries_count = binaries_count;
 	for (int i = 0; i < iproto_threads_count; i++)
 		if (iproto_do_cfg(&iproto_threads[i], &cfg_msg) != 0)
 			return -1;
 	return 0;
 }
 
 int
-iproto_listen(const char *uri)
+iproto_listen(const char **uri_array, uint32_t size)
 {
-	if (iproto_send_stop_msg() != 0)
+	assert(size <= IPROTO_LISTEN_SOCKET_MAX);
+	if (iproto_stop_listen() != 0)
 		return -1;
-	evio_service_stop(&tx_binary);
-	if (uri == NULL) {
-		tx_binary.addr_len = 0;
-		return 0;
+
+	uint32_t i;
+	for (i = 0; i < size; i++) {
+		/*
+		 * Please note, we bind socket in main thread, and then
+		 * listen this socket in all iproto threads! With this
+		 * implementation, we rely on the Linux kernel to distribute
+		 * incoming connections across iproto threads.
+		 */
+		if (evio_service_bind(&tx_binaries[i], uri_array[i]) != 0)
+			goto error;
 	}
-	/*
-	 * Please note, we bind socket in main thread, and then
-	 * listen this socket in all iproto threads! With this
-	 * implementation, we rely on the Linux kernel to distribute
-	 * incoming connections across iproto threads.
-	 */
-	if (evio_service_bind(&tx_binary, uri) != 0)
-		return -1;
-	if (iproto_send_listen_msg(&tx_binary) != 0)
+	if (iproto_send_listen_msg(tx_binaries, size) != 0)
+		goto error;
+
+	tx_binaries_count = size;
+	return 0;
+error:
+	tx_binaries_count = i;
```

Comment:
Nit: the less number of functions changing global vars the easier it turns out to read the code. Personally I would nullify tx_binaries_count right here and pass i to iproto_stop_listen. But it's my taste.

Reply:
fixed